src/System.Management.Automation/engine/CoreAdapter.cs

```diff
@@ -1693,13 +1696,19 @@ internal static Type EffectiveArgumentType(object arg)
                         return firstType.MakeArrayType();
                     }
                 }
-
-                return arg.GetType();
             }
-            else
+
+            return GetObjectType(arg, debase: false);
+        }
+
+        internal static Type GetObjectType(object obj, bool debase)
+        {
+            if (debase)
             {
-                return typeof(LanguagePrimitives.Null);
+                obj = PSObject.Base(obj);
             }
+
+            return obj == NullString.Value ? typeof(string) : obj?.GetType();
```

Comment:
Is it really hot path code? If so we could think about more tier-compilation-friendly methods:
```c#
        internal static Type GetObjectTypeAfterDebase(object obj)
        {
            obj = PSObject.Base(obj);

            return obj == NullString.Value ? typeof(string) : obj?.GetType();
    }

        internal static Type GetObjectTypeWithoutDebase(object obj) => 
obj == NullString.Value ? typeof(string) : obj?.GetType();

Reply:
I think I know C# pretty well but never considered it from performance side. In C++, this function would be almost guaranteed to be inlined and the `if` discarded or made unconditional. Is C#'s optimizer really not capable of it? Do you have some optimization guidelines I could read, e.g. when to use `[AggressivelyInline]` I saw in some `WriteToConsole` overloads?

Reply:
For reading https://devblogs.microsoft.com/dotnet/performance_improvements_in_net_7/

Reply:
This seems to be a micro-optimization that we should avoid, as I don't think the difference is measurable.